multiplexer/abci/remote_v1.go

```diff
@@ -0,0 +1,636 @@
+package abci
+
+import (
+	"context"
+	"math"
+
+	"google.golang.org/grpc"
+
+	abciv2 "github.com/cometbft/cometbft/abci/types"
+	cryptov2 "github.com/cometbft/cometbft/proto/tendermint/crypto"
+	typesv2 "github.com/cometbft/cometbft/proto/tendermint/types"
+	abciv1 "github.com/tendermint/tendermint/abci/types"
+	cryptov1 "github.com/tendermint/tendermint/proto/tendermint/crypto"
+	typesv1 "github.com/tendermint/tendermint/proto/tendermint/types"
+	versionv1 "github.com/tendermint/tendermint/proto/tendermint/version"
+)
+
+type RemoteABCIClientV1 struct {
+	abciv1.ABCIApplicationClient
+
+	// retainLastHeight is the height is set in finalize block
+	// and returned in commit
+	commitRetainLastHeight int64
+	// endBlockConsensusVersion is the app version got from the end block abci call
+	endBlockConsensusAppVersion uint64
+	// chainID is required to pass into the headers.
+	chainID string
+}
+
+// NewRemoteABCIClientV1 returns a new ABCI Client (using ABCI v1).
+// The client behaves like Tendermint for the server side (the application side).
+func NewRemoteABCIClientV1(conn *grpc.ClientConn, chainID string) *RemoteABCIClientV1 {
+	return &RemoteABCIClientV1{
+		ABCIApplicationClient: abciv1.NewABCIApplicationClient(conn),
+		chainID:               chainID,
+	}
+}
+
+// ApplySnapshotChunk implements abciv2.ABCI
+func (a *RemoteABCIClientV1) ApplySnapshotChunk(req *abciv2.RequestApplySnapshotChunk) (*abciv2.ResponseApplySnapshotChunk, error) {
+	resp, err := a.ABCIApplicationClient.ApplySnapshotChunk(
+		context.Background(),
+		&abciv1.RequestApplySnapshotChunk{
+			Index:  req.Index,
+			Chunk:  req.Chunk,
+			Sender: req.Sender,
+		},
+		grpc.WaitForReady(true),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &abciv2.ResponseApplySnapshotChunk{
+		Result:        abciv2.ResponseApplySnapshotChunk_Result(resp.Result),
+		RefetchChunks: resp.RefetchChunks,
+		RejectSenders: resp.RejectSenders,
+	}, nil
+}
+
+// CheckTx implements abciv2.ABCI
+func (a *RemoteABCIClientV1) CheckTx(req *abciv2.RequestCheckTx) (*abciv2.ResponseCheckTx, error) {
+	resp, err := a.ABCIApplicationClient.CheckTx(context.Background(), &abciv1.RequestCheckTx{
+		Tx:   req.Tx,
+		Type: abciv1.CheckTxType(req.Type),
+	}, grpc.WaitForReady(true))
+	if err != nil {
+		return nil, err
+	}
+
+	return &abciv2.ResponseCheckTx{
+		Code:      resp.Code,
+		Data:      resp.Data,
+		Log:       resp.Log,
+		Info:      resp.Info,
+		GasWanted: resp.GasWanted,
+		GasUsed:   resp.GasUsed,
+		Events:    abciEventV1ToV2(resp.Events...),
+		Codespace: resp.Codespace,
+	}, nil
+}
+
+// Commit implements abciv2.ABCI
+func (a *RemoteABCIClientV1) Commit() (*abciv2.ResponseCommit, error) {
+	return &abciv2.ResponseCommit{
+		RetainHeight: a.commitRetainLastHeight,
+	}, nil
+}
+
+// FinalizeBlock implements abciv2.ABCI
+func (a *RemoteABCIClientV1) FinalizeBlock(req *abciv2.RequestFinalizeBlock) (*abciv2.ResponseFinalizeBlock, error) {
```

Comment:
I would've expected ABCI v1 to support EndBlock, DeliverTx and not support FinalizeBlock

Reply:
So the app (v3) is using ABCI 1, but consensus is always ABCI 2. In remote v1 we basically map consensus abci calls to the proper abci v1 flow for the app.